File: 4Periodo/EstruturasDeDados/1Avaliacao/testeEstruturasDeDados/teste_fila.c
#include <stdbool.h>
#include <stdlib.h>
typedef struct {
  int primeiro;
  int ultimo;
  int *estrutura;
  int maxItens;
} Fila;
void new_fila(Fila *f, int capacidade) {
  f->estrutura = (int *)malloc(capacidade * sizeof(int));
  f->primeiro = 0;
  f->ultimo = 0;
  f->maxItens = capacidade;
}
bool esta_cheia(Fila *f) {
  return ((f->ultimo + 1) % f->maxItens) == f->primeiro;
}
bool esta_vazia(Fila *f) { return f->primeiro == f->ultimo; }
bool push(Fila *f, int novo_item) {
  if (esta_cheia(f)) {
    return false;
  }
  f->estrutura[f->ultimo] = novo_item;
  f->ultimo = (f->ultimo + 1) % f->maxItens;
  return true;
}
bool pop(Fila *f, int *valor) {
  if (esta_vazia(f)) {
    return false;
  }
  *valor = f->estrutura[f->primeiro];
  f->primeiro = (f->primeiro + 1) % f->maxItens;
  return true;
}
bool view(Fila *f, int *valor) {
  if (esta_vazia(f)) {
    return false;
  }
  *valor = f->estrutura[f->primeiro];

  return true;
}

File: 4Periodo/EstruturasDeDados/1Avaliacao/testeEstruturasDeDados/teste_fila.zig
const std = @import("std");
pub const ChanError = error{
    EstaCheia,
    EstaVazia,
};
const Fila = struct {
    primeiro: usize,
    ultimo: usize,
    estrutura: []usize,
    maxItens: i32,
    pub fn init(primeiro: usize, estrutura: []i32, maxItens: i32) Fila {
        return Fila{
            .primeiro = primeiro,
            .ultimo = primeiro, //Ao inicializar fila, primeiro é o ultimo também.
            .estrutura = estrutura,
            .maxItens = maxItens,
        };
    }
    fn estaCheia(self: Fila) bool {
        return ((self.ultimo + 1) % self.maxItens) == self.primeiro;
    }
    fn estaVazia(self: Fila) bool {
        return self.primeiro == self.ultimo;
    }
    fn push(self: *Fila, novoItem: usize) ChanError!void {
        if (self.estaCheia()) {
            return ChanError.EstaCheia;
        }
        self.estrutura[self.ultimo] = novoItem;
        self.ultimo = (self.ultimo + 1) % self.maxItens;
        return;
    }
    fn pop(self: *Fila) ChanError!usize {
        if (self.estaVazia()) {
            return ChanError.EstaVazia;
        }
        const valor: usize = self.estrutura[self.primeiro];
        self.primeiro = (self.primeiro + 1) % self.maxItens;
        return valor;
    }
    fn view(self: Fila) ChanError!usize {
        if (self.estaVazia()) {
            return ChanError.EstaVazia;
        }
        return self.estrutura[self.primeiro];
    }
};
